### PR TITLE
Merge topological sort code to eMPRess

### LINKED ISSUE #37: Merge topological ordering code to this repository
Merge topological ordering (Rose) code from [JingyiRose/TopologicalSort](https://github.com/JingyiRose/TopologicalSort) to this repository.

### PR DESCRIPTION
**Pull request by JingyiRose**
Add `Tree` class, `tree_format_converter.py` file that converts a parasite or host tree from dictionary format to a `Tree` structure,  and a `recon_builder.py` file that provides `build_trees_with_temporal_order` function which uses topological sort to populate the order of nodes inside host and parasite tree and output two `Tree`s that can be used for visualization.

Resolves #37.
This pull request is linked to https://github.com/JingyiRose/TopologicalSort/pull/1

### CODE REVIEW

File: topoSort/ReconBuilder.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+from Tree import NodeLayout
+from Tree import TreeType
+from TreeFormatConverter import dict_to_tree
+
+__all__ = ["build_trees_with_temporal_order", "build_temporal_graph", "topological_order"]
+
+def build_trees_with_temporal_order(host_tree, parasite_tree, reconciliation):
+    """
+    :param host_tree: host tree dictionary
```
Comment (ssantichaivekin) [inline topoSort/ReconBuilder.py:14]: I think you should add some overview docstring comment before the first `:param`.

File: topoSort/ReconBuilder.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
+# ReconBuilder.py
```
Comment (ssantichaivekin) [inline topoSort/ReconBuilder.py:1]: If you have time, please convert most of the dicts and trees to classes.

File: topoSort/ReconBuilderTest.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
+import ReconBuilder
```
Comment (ssantichaivekin) [inline topoSort/ReconBuilderTest.py:1]: can you convert this to use `unittest`?

File: topoSort/ReconBuilder.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+# This code builds a representation of a host tree, parasite tree, and a reconfiguration
+# as specified in Recon.py using the host tree, parasite tree, and reconfiguration 
+# representation in the DTL.
+
+from Tree import NodeLayout
+from Tree import TreeType
+from TreeFormatConverter import dict_to_tree
+
+__all__ = ["build_trees_with_temporal_order", "build_temporal_graph", "topological_order"]
```
Comment (ssantichaivekin) [inline topoSort/ReconBuilder.py:10]: I still think you should only export `build_trees_with_temporal_order`.

File: topoSort/ReconBuilderTest.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+('n8', 'm3'): [('C', (None, None), (None, None))],
+('n5', 'm2'): [('C', (None, None), (None, None))],
+('n6', 'm3'): [('C', (None, None), (None, None))]}
+
+def check_topological_order(temporal_graph, ordering_dict):
+  """ Check the ordering_dict orders the nodes in temporal_graph in topological order"""
+  
+  for node_tuple in temporal_graph:
+    for node_tuple_child in temporal_graph[node_tuple]:
+      is_child_leaf = bool(node_tuple_child not in ordering_dict)
```
Comment (ssantichaivekin) [inline topoSort/ReconBuilderTest.py:106]: no need for the `bool`.

File: topoSort/ReconBuilderTest.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+      # leaves are not ordered
+      if is_child_leaf:
+        continue
+      assert(ordering_dict[node_tuple] < ordering_dict[node_tuple_child])
+
+def print_order_in_tree(tree_node):
```
Comment (ssantichaivekin) [inline topoSort/ReconBuilderTest.py:112]: remove this from testing

**Comment by ssantichaivekin** [comment]
Instead of creating a folder at `topoSort` (actually this should be snake case `topo_sort`), create a folder at `empress/topo_sort`.

**Comment by ssantichaivekin** [comment]
You might want to integrate new changes from branch `topoSort2` written by Prof Ran.

File: topoSort/ReconBuilder.py (hunk 4 above)
Comment (JingyiRose) [reply topoSort/ReconBuilder.py:10]: I think all three functions are useful. I might use the latter two functions myself, so I think I'll keep it there for now.

File: topoSort/ReconBuilder.py (hunk 2 above)
Comment (JingyiRose) [reply topoSort/ReconBuilder.py:1]: will add to future tasks

File: topoSort/ReconBuilder.py (hunk 1 above)
Comment (JingyiRose) [reply topoSort/ReconBuilder.py:14]: will do

File: topoSort/ReconBuilderTest.py (hunk 3 above)
Comment (JingyiRose) [reply topoSort/ReconBuilderTest.py:1]: will do

File: topoSort/ReconBuilderTest.py (hunk 5 above)
Comment (JingyiRose) [reply topoSort/ReconBuilderTest.py:106]: will do

File: topoSort/ReconBuilderTest.py (hunk 6 above)
Comment (JingyiRose) [reply topoSort/ReconBuilderTest.py:112]: will do

File: TreeTester.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
+# TreeTester.py
```
Comment (ssantichaivekin) [inline TreeTester.py:1]: remove this

File: TreeFormatConverter.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
+# TreeFormatConverter.py
```
Comment (ssantichaivekin) [inline TreeFormatConverter.py:1]: remove this

File: Tree.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
+# Tree.py
```
Comment (ssantichaivekin) [inline Tree.py:1]: remove this

File: empress/topoSort/ReconBuilder.py [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+        for node in ordering_dict:
+            if max_order < ordering_dict[node]:
+                max_order = ordering_dict[node]
+        leaf_order = max_order + 1
+        populate_nodes_with_order(host_tree_object.root_node, TreeType.HOST, ordering_dict, leaf_order)
+        populate_nodes_with_order(parasite_tree_object.root_node, TreeType.PARASITE, ordering_dict, leaf_order)
+
+    return host_tree_object, parasite_tree_object
+
+def create_parent_dict(H, P):
```
Comment (ssantichaivekin) [inline empress/topoSort/ReconBuilder.py:42]: Use `host_tree` instead of `H`

File: empress/topoSort/ReconBuilder.py [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+            if child_node in unvisited_nodes:
+                unvisited_nodes.remove(child_node)
+            has_cycle_child, next_order = topological_order_helper(child_node, next_order,  visiting_nodes,
+                                   unvisited_nodes, temporal_graph, ordering_dict)
+            # if we find a cycle, we stop the process
+            if has_cycle_child: return True, next_order
+        # if children are all labeled, we can label the start_node
+        visiting_nodes.remove(start_node)
+        ordering_dict[start_node] = next_order
+        return False, next_order+1
```
Comment (ssantichaivekin) [inline empress/topoSort/ReconBuilder.py:217]: `next_order + 1`. `(next_order+1) * 5`

File: empress/topoSort/Tree.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+    def _postorder_list_helper(self, node):
+        if node.is_leaf: return [node]
+        else:
+            list1 = self._postorder_list_helper(node.left_node)
+            list2 = self._postorder_list_helper(node.right_node)
+            list1.extend(list2)
+            list1.append(node)
+            return list1
+
```
Comment (ssantichaivekin) [inline empress/topoSort/Tree.py:77]: remove extra spaces.

File: empress/topoSort/test_ReconBuilder.py [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+        """
+        Check the ordering_dict orders the nodes in temporal_graph in topological order
+        """
+        for node_tuple in temporal_graph:
+            for node_tuple_child in temporal_graph[node_tuple]:
+                is_child_leaf = node_tuple_child not in ordering_dict
+                # leaves are not ordered
+                if is_child_leaf:
+                    continue
+                assert(ordering_dict[node_tuple] < ordering_dict[node_tuple_child])
```
Comment (ssantichaivekin) [inline empress/topoSort/test_ReconBuilder.py:44]: use `self.assertTrue(...)`.

File: empress/topoSort/test_ReconBuilder.py [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+        ('n1', 'm1'): [('L', ('n1', 'm2'), (None, None))],
+        ('n1', 'm2'): [('T', ('n2', 'm2'), ('n3', 'm4'))],
+        ('n2', 'm2'): [('C', (None, None), (None, None))],
+        ('n4', 'm5'): [('C', (None, None), (None, None))],
+        ('n5', 'm5'): [('C', (None, None), (None, None))],
+        ('n6', 'm6'): [('C', (None, None), (None, None))],
+        ('n3', 'm4'): [('S', ('n4', 'm5'), ('n6', 'm6'))]}
+
+        ordering_dict = TestReconBuilder.build_ordering_dict(host_tree, parasite_tree, reconciliation)
+        assert(ordering_dict == None)
```
Comment (ssantichaivekin) [inline empress/topoSort/test_ReconBuilder.py:78]: use `self.assertTrue(...)`. Is there a `self.assertNone` or something like that? You should check :)

File: empress/topoSort/test_ReconBuilder.py [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+        reconciliation = { ('n0', 'm4'): [('T', ('n4', 'm4'), ('n2', 'm2'))],
+        ('n2', 'm2'): [('T', ('n5', 'm2'), ('n3', 'm1'))],
+        ('n3', 'm1'): [('S', ('n1', 'm3'), ('n6', 'm4'))],
+        ('n1', 'm3'): [('C', (None, None), (None, None))],
+        ('n4', 'm4'): [('C', (None, None), (None, None))],
+        ('n5', 'm2'): [('C', (None, None), (None, None))],
+        ('n6', 'm4'): [('C', (None, None), (None, None))]}
+
+        ordering_dict = TestReconBuilder.build_ordering_dict(host_tree, parasite_tree, reconciliation)
+        assert(ordering_dict == None)
```
Comment (ssantichaivekin) [inline empress/topoSort/test_ReconBuilder.py:109]: `self.assertIsNone(...)`

File: empress/topoSort/test_ReconBuilder.py [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+                        if ordering_dict != None:
+                            TestReconBuilder.check_topological_order(temporal_graph, ordering_dict)
+                count += 1
+
+    @classmethod
+    def tearDown(self):
+        """
+        clean up the generated tests
+        """
+        generated_dir_path = "./newickSample"
```
Comment (ssantichaivekin) [inline empress/topoSort/test_ReconBuilder.py:144]: Put it as class variable. (`generated_dir_path` should not be declared twice once for `setUp` and another for `tearDown`).

File: empress/topoSort/test_ReconBuilder.py [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+    """
+    Test the build_trees_with_temporal_order function from ReconBuilder
+    """
+    @classmethod
+    def setUp(self):
+        """
+        generate newick tests of certain sizes
+        """
+        size_range = [5]
+        generated_dir_path = "./newickSample"
```
Comment (ssantichaivekin) [inline empress/topoSort/test_ReconBuilder.py:21]: I prefer `./temp_newick_samples` as the directory path. The `temp` is there to ensure the name does not match with other folders in the future.

File: empress/topoSort/test_ReconBuilder.py [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+
+    @classmethod
+    def test_topological_order(self):
+        """
+        test topological_order by generating host and parasite trees of different sizes and going through
+        each reconciliaiton in the reconciliaiton graph to see if the order generated by topological_order
+        is a topological order
+        """
+        size_range = [5]
+        generated_dir_path = "./newickSample"
```
Comment (ssantichaivekin) [inline empress/topoSort/test_ReconBuilder.py:119]: make this a class variable and don't redefine it every time.

File: empress/topoSort/test_ReconBuilder.py [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+        assert(ordering_dict == None)
+
+    @classmethod
+    def test_topological_order(self):
+        """
+        test topological_order by generating host and parasite trees of different sizes and going through
+        each reconciliaiton in the reconciliaiton graph to see if the order generated by topological_order
+        is a topological order
+        """
+        size_range = [5]
```
Comment (ssantichaivekin) [inline empress/topoSort/test_ReconBuilder.py:118]: make this a class variable and don't redefine it every time.

File: empress/topoSort/test_ReconBuilder.py [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+        generated_dir_path = "./newickSample"
+        count = 0
+        for tree_size in size_range:
+            tree_size_Folder = '%s/size%d' % (generated_dir_path, tree_size)
+            for newick in os.listdir(tree_size_Folder):
+                if count >= 5: break # we only test 5 examples
```
Comment (ssantichaivekin) [inline empress/topoSort/test_ReconBuilder.py:124]: Make this a class variable too.

File: empress/topoSort/test_ReconBuilder.py [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+        ('n5', 'm2'): [('C', (None, None), (None, None))],
+        ('n6', 'm4'): [('C', (None, None), (None, None))]}
+
+        ordering_dict = TestReconBuilder.build_ordering_dict(host_tree, parasite_tree, reconciliation)
+        assert(ordering_dict == None)
+
+    @classmethod
+    def test_topological_order(self):
+        """
+        test topological_order by generating host and parasite trees of different sizes and going through
```
Comment (ssantichaivekin) [inline empress/topoSort/test_ReconBuilder.py:114]: Write a comment that says that most trees in this test is temporally consistent.

File: empress/topoSort/test_ReconBuilder.py [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+import itertools
+import shutil
+from empress import newickFormatReader
+from empress.topoSort import ReconBuilder
+from empress.clumpr import DTLReconGraph
+from empress.clumpr import HistogramAlgTools
+from empress.clumpr.script02_gen_newick_trees import generateNewickTestsMultipleSizes
+
+class TestReconBuilder(unittest.TestCase):
+    """
```
Comment (ssantichaivekin) [inline empress/topoSort/test_ReconBuilder.py:12]: Write a comment that says that most trees in the bottom test is temporally consistent, while the inconsistent trees are tested separately.

File: empress/topoSort/test_ReconBuilder.py (hunk 15 above)
Comment (JingyiRose) [reply empress/topoSort/test_ReconBuilder.py:109]: I will use assertFalse

File: empress/topoSort/test_ReconBuilder.py (hunk 21 above)
Comment (ssantichaivekin) [reply empress/topoSort/test_ReconBuilder.py:200]: Start a block comment with capital letter. "Test topological_order ..."

File: empress/topoSort/test_ReconBuilder.py [hunk 23]
```diff
@@ -0,0 +1,231 @@
...
+        ordering_dict = ReconBuilder.topological_order(temporal_graph)
+        self.assertIsNone(ordering_dict)
+
+        host, parasite, if_consistent = ReconBuilder.build_trees_with_temporal_order(host_tree,
+                                                                    parasite_tree, reconciliation)
+        self.assertIsNone(host)
+        self.assertIsNone(parasite)
+        self.assertFalse(if_consistent)
+
+    def test_detect_consistency_1(self):
```
Comment (ssantichaivekin) [inline empress/topoSort/test_ReconBuilder.py:76]: Nice :). Thanks for doing this :)

File: empress/topoSort/test_ReconBuilder.py [hunk 24]
```diff
@@ -0,0 +1,231 @@
...
+    examples that might be temporally consistent or inconsistent.
+    """
+
+    size_range = [5] # range of sizes of the tree examples we want to generate
+    generated_dir_path = "./temp_newick_samples" # directory for generated examples
+    num_examples_to_test = 5 # number of examples to run in the last test
+
+    def setUp(self):
+        """
+        generate newick tests of certain sizes
```
Comment (ssantichaivekin) [inline empress/topoSort/test_ReconBuilder.py:25]: Starts block comment with capital letter.

File: empress/topoSort/TreeTester.py [hunk 25]
```diff
@@ -0,0 +1,42 @@
+# TreeTester.py
```
Comment (ssantichaivekin) [inline empress/topoSort/TreeTester.py:1]: Move to `test/topoSort/test_Tree.py` (create a new folder called topoSort in the folder test and move this there).

File: empress/topoSort/test_ReconBuilder.py [hunk 26]
```diff
@@ -0,0 +1,231 @@
+import unittest
```
Comment (ssantichaivekin) [inline empress/topoSort/test_ReconBuilder.py:1]: Move to `test/topoSort/test_ReconBuilder.py` (create a new folder called topoSort in the folder test and move this there).

File: empress/topoSort/ReconBuilder.py [hunk 27]
```diff
@@ -0,0 +1,277 @@
+# ReconBuilder.py
```
Comment (ssantichaivekin) [inline empress/topoSort/ReconBuilder.py:1]: Do you want me to rename `topoSort` to `topo_sort` and `ReconBuilder` to `recon_builder` for you?

File: empress/topoSort/test_ReconBuilder.py [hunk 28]
```diff
@@ -0,0 +1,231 @@
...
+                        # if there is no temporal inconsistency
+                        if ordering_dict != None:
+                            self.check_topological_order(temporal_graph, ordering_dict)
+                count += 1
+
+    def tearDown(self):
+        """
+        clean up the generated tests
+        """
+        if os.path.isdir(self.generated_dir_path):
```
Comment (ssantichaivekin) [inline empress/topoSort/test_ReconBuilder.py:227]: You shouldn't have to check if there is a directory there.

File: empress/topoSort/Tree.py [hunk 29]
```diff
@@ -0,0 +1,72 @@
...
+        if node.is_leaf: return [node]
+        else:
+            list1 = self._leaf_list_helper(node.left_node)
+            list2 = self._leaf_list_helper(node.right_node)
+            list1.extend(list2)
+            return list1
+
+
+    def postorder_list(self):
+        """ returns list of all Nodes in postorder """
```
Comment (ssantichaivekin) [inline empress/topoSort/Tree.py:62]: Capital letter at the start of block comment

File: empress/topoSort/ReconBuilder.py [hunk 30]
```diff
@@ -0,0 +1,277 @@
...
+
+from empress.topoSort.Tree import NodeLayout
+from empress.topoSort.Tree import TreeType
+from empress.topoSort.TreeFormatConverter import dict_to_tree
+
+__all__ = ["build_trees_with_temporal_order"]
+
+def build_trees_with_temporal_order(host_tree, parasite_tree, reconciliation):
+    """
+    This is the main function for visualization purposes.
```
Comment (ssantichaivekin) [inline empress/topoSort/ReconBuilder.py:14]: I think you might want to change this comment to "Use topological sort to order the vertices inside host and parasite tree. The output trees can be used for visualization."

File: empress/topoSort/test_ReconBuilder.py (hunk 28 above)
Comment (JingyiRose) [reply empress/topoSort/test_ReconBuilder.py:227]: you are correct, tearDown is only called if setup succeeds

File: empress/topoSort/ReconBuilder.py (hunk 27 above)
Comment (JingyiRose) [reply empress/topoSort/ReconBuilder.py:1]: I can do it, thanks for suggesting

File: test/topo_sort/test_ReconBuilder.py [hunk 31]
```diff
@@ -0,0 +1,230 @@
+import unittest
```
Comment (ssantichaivekin) [inline test/topo_sort/test_ReconBuilder.py:1]: Use `snake_case`.

File: empress/topo_sort/TreeFormatConverter.py [hunk 32]
```diff
@@ -0,0 +1,50 @@
+# TreeFormatConverter.py
```
Comment (ssantichaivekin) [inline empress/topo_sort/TreeFormatConverter.py:1]: Use `snake_case`.

**Pull request merged**